File: Lession 3/bfs.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
int bfs(int u, vector<vector<int>> &graph, vector<bool> &visited)
{
    queue<int> q;
    q.push(u);
    visited[u] = true;
    while (!q.empty())
    {
        int u = q.front();
        q.pop();
        for (int j = 0; j < graph[u].size(); j++)
        {
            int v = graph[u][j];
            if (!visited[v])
            {
                q.push(v);
                visited[v] = true;
            }
        }
    }
}
int main()
{
    int n, m;
    cin >> n >> m;
    vector<vector<int>> graph(n);
    vector<bool> visited(n, false);
    for (int i = 0; i < m; i++)
    {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    int count = 0;
    for (int i = 0; i < n; i++)
    {
        if (!visited[i])
        {
            bfs(i, graph, visited);
            count++;
        }
    }
    cout << count << endl;
    return 0;
}

File: Lession 3/bfs.jl
function bfs(graph, start)
    visited = Dict{Int, Bool}()
    for node in keys(graph)
        visited[node] = false
    end
    queue = [start]
    visited[start] = true
    while length(queue) > 0
        node = popfirst!(queue)
        println(node)
        for neighbor in graph[node]
            if !visited[neighbor]
                push!(queue, neighbor)
                visited[neighbor] = true
            end
        end
    end
end
function main()
    graph = Dict{Int, Vector{Int}}()
    graph[1] = [2, 3]
    graph[2] = [4]
    graph[3] = [5]
    graph[4] = [6]
    graph[5] = [6]
    graph[6] = []
    bfs(graph, 1)
end